habitat/tasks/nav/greedy_follower.py

```diff
@@ -0,0 +1,210 @@
+            axis = vh.T[:, 2]
+            w2 = (1 + c) * 0.5
+            w = np.sqrt(w2)
+            axis = axis * np.sqrt(1 - w2)
+            return np.quaternion(w, *axis)
+
+        axis = np.cross(v0, v1)
+        s = np.sqrt((1 + c) * 2)
+        return np.quaternion(s * 0.5, *(axis / s))
+
+    def _geo_dist(self, goal_pos: np.array) -> float:
+        return self._sim.geodesic_distance(
+            self._sim.get_agent_state().position, goal_pos
+        )
+
+    def _est_max_grad_dir(self, goal_pos: np.array) -> np.array:
+
+        current_pos = self._sim.get_agent_state().position
+
+        if self.mode == "spath":
+            points = self._sim.straight_spath_points(
+                self._sim.get_agent_state().position, goal_pos
+            )
+            # NB: Add a little offset as things get weird if
+            # points[1] - points[0] is anti-parallel with forward
+            if len(points) < 2:
+                return None
+            max_grad_dir = self._quat_from_two_vectors(
+                self._sim.FORWARD,
+                points[1]
+                - points[0]
+                + 1e-3 * np.cross(self._sim.UP, self._sim.FORWARD),
+            )
+        else:
+            current_rotation = self._sim.get_agent_state().rotation
+            current_dist = self._geo_dist(goal_pos)
+
+            best_delta = -1e10
+            best_rotation = current_rotation
+            for _ in range(0, 360, self._sim.config.TURN_ANGLE):
+                sim_action = SIM_NAME_TO_ACTION[SimulatorActions.FORWARD.value]
+                self._sim.step(sim_action)
+                new_delta = current_dist - self._geo_dist(goal_pos)
+
+                if new_delta > best_delta:
+                    best_rotation = self._sim.get_agent_state().rotation
+                    best_delta = new_delta
+
+                # If the best delta is within (1 - cos(TURN_ANGLE))% of the best
+                # delta (the step size), then we almost certianly have found the
+                # max grad dir and should just exit
+                if np.isclose(
+                    best_delta,
+                    self._max_delta,
+                    rtol=1 - np.cos(np.deg2rad(self._sim.config.TURN_ANGLE)),
+                ):
+                    break
+
+                self._sim.set_agent_state(
+                    current_pos,
+                    self._sim.get_agent_state().rotation,
+                    reset_sensors=False,
+                )
+
+                sim_action = SIM_NAME_TO_ACTION[SimulatorActions.LEFT.value]
+                self._sim.step(sim_action)
+
+            self._sim.set_agent_state(
+                current_pos, current_rotation, reset_sensors=False
+            )
+
+            max_grad_dir = v_to_q(best_rotation)
+
+        return max_grad_dir
+
+    @property
+    def mode(self):
+        return self._mode
+
+    @mode.setter
+    def mode(self, new_mode):
+        assert new_mode in {"spath", "grad"}
```

Comment:
That would be great to add more description what "spath", "grad" stands here.

Reply:
Done.